### PR TITLE
Functional tests rewrite - Type 0 - #678

### PR DESCRIPTION
**Pull request by diego-G**
First part of functional tests rewrite for transaction type 0.

Subtask of #678

### CODE REVIEW

File: test/functional/http/post/0.tx.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+	describe('schema', function () {
+
+		var transaction;
+
+		beforeEach(function () {
+			transaction = node.randomTx();
+		});
+
+		describe('sending transactions without proper format', function () {
```
Comment (karmacoma) [inline test/functional/http/post/0.tx.js]: I don't think we need this extra context.

This would be fine:

```
POST /api/transactions (type 0)
  schema checks
     using null should fail
 ```

File: test/functional/ws/transport.transactions.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+var shared = require('../http/post/shared');
+var ws = require('../../common/wsCommunication');
+
+function postTransaction (transaction, done) {
+	ws.call('postTransactions', {
+		transaction: transaction
+	}, done, true);
+}
+
+describe('Basic Transactions via websockets', function () {
```
Comment (karmacoma) [inline test/functional/ws/transport.transactions.js]: `Posting transaction (type 0)`

For other `ws/transport` tests, websockets is implied.

File: Jenkinsfile [hunk 3]
```diff
@@ -128,111 +128,120 @@ lock(resource: "Lisk-Core-Nodes", inversePrecedence: true) {
...
-					sh '''
-					cd "$(echo $WORKSPACE | cut -f 1 -d '@')"
-					npm run eslint
-					'''
-				}
+						sh '''
+						cd "$(echo $WORKSPACE | cut -f 1 -d '@')"
+						npm run eslint
+						'''
+					}
```
Comment (Isabello) [inline Jenkinsfile]: This formatting is wrong, there are too many tabs.

File: package.json [hunk 4]
```diff
@@ -9,9 +9,9 @@
     "fetchCoverage": "./node_modules/.bin/grunt exec:fetchCoverage --verbose",
     "coverageReport": "./node_modules/.bin/grunt coverageReport --verbose",
     "test": "./node_modules/.bin/grunt test --verbose",
-    "test-unit": "./node_modules/.bin/grunt test-unit --verbose",
-    "test-integration": "export TEST=test/integration/peers.integration.js && export NODE_ENV=TEST && ./node_modules/.bin/grunt jenkins",
-    "test-functional": "grunt test-functional",
+    "test-unit": "./node_modules/.bin/grunt testUnit --verbose",
+    "test-functional": "./node_modules/.bin/grunt testFunctional --verbose",
+    "test-integration": "./node_modules/.bin/grunt testIntegration",
```
Comment (Isabello) [inline package.json]: Why is this being performed as part of this PR? this should be a separate PR.

File: Gruntfile.js [hunk 5]
```diff
@@ -141,8 +146,9 @@ module.exports = function (grunt) {
 	grunt.registerTask('coverageReport', ['exec:coverageReport']);
 	grunt.registerTask('eslint-nofix', ['eslint']);
 	grunt.registerTask('test', ['eslint', 'exec:coverage']);
-	grunt.registerTask('test-unit', ['eslint', 'exec:coverageUnit']);
-	grunt.registerTask('test-functional', ['eslint', 'exec:testFunctional']);
+	grunt.registerTask('testUnit', ['eslint', 'exec:testUnit']);
+	grunt.registerTask('testFunctional', ['eslint', 'exec:testFunctional']);
```
Comment (Isabello) [inline Gruntfile.js]: Why is this being performed as part of this PR? this should be a separate PR.

File: test/functional/http/get/transactions.js [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
-var sendLiskTrs = require('../common/complexTransactions.js').sendLISK;
-var transactionSortFields = require('../../sql/transactions').sortFields;
-var modulesLoader = require('../common/initModule').modulesLoader;
-var transactionTypes = require('../../helpers/transactionTypes.js');
-var genesisblock = require('../genesisBlock.json');
+var node = require('../../../node.js');
+var http = require('../../../common/httpCommunication.js');
+var ws = require('../../../common/wsCommunication');
+var sendLiskTrs = require('../../../common/complexTransactions.js').sendLISK;
+var transactionSortFields = require('../../../../sql/transactions').sortFields;
```
Comment (Isabello) [inline test/functional/http/get/transactions.js]: Why is this stuff being initialized as part of each module? We should be init'ing modules and functions inside of the `node.js` file at the top level of the test structure.

File: test/functional/http/post/0.tx.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
+'use strict';
+
+var node = require('../../../node');
+var shared = require('./shared');
```
Comment (Isabello) [inline test/functional/http/post/0.tx.js]: this is inconsistent. in some files we load by .js, in others we are loading without .js

Please stick to the standard format of loading with `.js` for anything that is not a module.

File: test/functional/http/post/0.tx.js [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+var constants = require('../../../../helpers/constants');
+
+var sendTransaction = require('../../../common/complexTransactions').sendTransaction;
+
+describe('POST /api/transactions (type 0)', function () {
+
+	var badTransactions = [];
+	var goodTransactions = [];
+
+	describe('schema checks', function () {
```
Comment (Isabello) [inline test/functional/http/post/0.tx.js]: schema validation

File: test/functional/http/post/0.tx.js [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+		it('using empty object should fail', function (done) {
+			sendTransaction({}, function (err, res) {
+				node.expect(res).to.have.property('success').not.to.be.ok;
+				node.expect(res).to.have.property('message').to.equal('Invalid transaction body - Empty trs passed');
+				done();
+			}, true);
+		});
+	});
+
+	describe('processing', function () {
```
Comment (Isabello) [inline test/functional/http/post/0.tx.js]: processing transactions

File: test/functional/http/post/0.tx.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+		it('sending transaction with same id twice should fail', function (done) {
+			sendTransaction(goodTransaction, function (err, res) {
+				node.expect(res).to.have.property('success').to.be.not.ok;
+				node.expect(res).to.have.property('message').to.equal('Transaction is already processed: '+goodTransaction.id);
+				done();
+			});
+		});
+	});
+
+	describe('confirmation', function () {
```
Comment (Isabello) [inline test/functional/http/post/0.tx.js]: transaction confirmations

File: test/functional/ws/transport.transactions.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+	var transaction;
+	var goodTransactions = [];
+	var badTransactions = [];
+	var account = node.randomAccount();
+
+	beforeEach(function () {
+		transaction = node.randomTx();
+	});
+
+	describe('processing', function () {
```
Comment (Isabello) [inline test/functional/ws/transport.transactions.js]: transaction processing

File: test/functional/ws/transport.transactions.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+			postTransaction(transaction, function (err, res) {
+				node.expect(res).to.have.property('success').to.be.ok;
+				node.expect(res).to.have.property('transactionId').to.equal(transaction.id);
+				goodTransactions.push(transaction);
+				done();
+			});
+		});
+	});
+
+	describe('confirmation', function () {
```
Comment (Isabello) [inline test/functional/ws/transport.transactions.js]: transaction confirmations

File: test/functional/http/post/0.tx.js (hunk 9 above)
Comment (karmacoma) [inline test/functional/http/post/0.tx.js]: transaction processing

Comment (Isabello) [inline test/functional/http/post/0.tx.js]: ok :)

File: Jenkinsfile (hunk 3 above)
Comment (diego-G) [inline Jenkinsfile]: I'm doing this but It should be caught by ESLINT.

File: package.json (hunk 4 above)
Comment (diego-G) [inline package.json]: Because it is a small change for development purposes and the discussion started thanks to this PR. @Isabello do really you want me to remove these changes and open a new issue/PR ?

File: Gruntfile.js (hunk 5 above)
Comment (diego-G) [inline Gruntfile.js]: Because it is a small change for development purposes and the discussion started thanks to this PR. @Isabello do really you want me to remove these changes and open a new issue/PR ?

File: test/functional/http/get/transactions.js (hunk 6 above)
Comment (diego-G) [inline test/functional/http/get/transactions.js]: We have #717 task list for exactly this purpose. I want to introduce the minimum unrelated changes in my PR.

File: test/functional/http/post/0.tx.js (hunk 7 above)
Comment (diego-G) [inline test/functional/http/post/0.tx.js]: @LiskHQ/backend what do we do here? with or without? I've changed this twice.

File: test/functional/http/post/0.tx.js (hunk 8 above)
Comment (diego-G) [inline test/functional/http/post/0.tx.js]: @karmacoma `schema checks` or  @Isabello  `schema validation`?

File: test/functional/http/post/0.tx.js (hunk 9 above)
Comment (diego-G) [inline test/functional/http/post/0.tx.js]: This is a new convention. Change the description of the issue if you prefer another way.

File: test/functional/http/post/0.tx.js (hunk 7 above)
Comment (MaciejBaj) [inline test/functional/http/post/0.tx.js]: I am a fan of not putting `.js` but I cannot find any articles to support me

File: test/functional/http/post/0.tx.js (hunk 8 above)
Comment (MaciejBaj) [inline test/functional/http/post/0.tx.js]: just schema :)

File: test/functional/http/post/0.tx.js (hunk 9 above)
Comment (karmacoma) [inline test/functional/http/post/0.tx.js]: Please go with Isabella's suggestions.

File: test/functional/http/post/0.tx.js (hunk 8 above)
Comment (karmacoma) [inline test/functional/http/post/0.tx.js]: Please go with Isabella's suggestions.

File: test/functional/http/post/0.tx.js (hunk 7 above)
Comment (karmacoma) [inline test/functional/http/post/0.tx.js]: For this PR, I would leave as is, i.e. without. Then we should open separate PR to standardise across whole application.

Comment (Isabello) [inline test/functional/http/post/0.tx.js]: Agree with @karmacoma on this one. We can wait and standardize later.

File: test/functional/http/get/transactions.js (hunk 6 above)
Comment (Isabello) [inline test/functional/http/get/transactions.js]: ok :)

File: test/functional/ws/transport.transactions.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	});
+
+	describe('transaction processing', function () {
+
+		it('when sender has no funds should fail', function (done) {
+			var transaction = node.lisk.transaction.createTransaction('1L', 1, account.password);
+
+			postTransaction(transaction, function (err, res) {
+				node.expect(res).to.have.property('success').to.be.not.ok;
+				node.expect(res).to.have.property('message').to.equal('Account does not have enough LSK: ' + account.address + ' balance: 0');
```
Comment (Isabello) [inline test/functional/ws/transport.transactions.js]: this has `+ var +` but other entries are `+var`.

We should make sure that we follow the same spacing standards everywhere. In this case the standard is right here but wrong elsewhere.

Here are some examples from the code base on the format.

https://github.com/LiskHQ/lisk/blob/development/logic/peers.js#L115
https://github.com/LiskHQ/lisk/blob/development/logic/transactionPool.js#L425
https://github.com/LiskHQ/lisk/blob/development/logic/transactionPool.js#L458

File: test/functional/http/post/0.tx.js [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+				node.expect(res).to.have.property('transactionId').to.equal(goodTransaction.id);
+				goodTransactions.push(goodTransaction);
+				done();
+			});
+		});
+
+		it('sending transaction with same id twice should fail', function (done) {
+			sendTransaction(goodTransaction, function (err, res) {
+				node.expect(res).to.have.property('success').to.be.not.ok;
+				node.expect(res).to.have.property('message').to.equal('Transaction is already processed: '+goodTransaction.id);
```
Comment (Isabello) [inline test/functional/http/post/0.tx.js]: see the comment about `+ var` spacing

**Comment by diego-G** [comment]
Due to the need of a better test suit structure, we close this issue and await for the restructure task presented in #717
